llmfoundry/callbacks/kill_loss_spike_callback.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Track training runs for loss spikes or persistently high training loss."""
+from __future__ import annotations
+
+import logging
+from collections import deque
+
+import numpy as np
+import torch
+from composer.core import Callback, State, TimeUnit
+from composer.loggers import Logger, MosaicMLLogger
+from composer.utils import dist
+
+from llmfoundry.utils.exceptions import HighLossError, LossSpikeError
+from llmfoundry.utils.warnings import experimental_class
+
+log = logging.getLogger(__name__)
+
+__all__ = ['KillLossSpike']
+
+MIN_WINDOW_SIZE = 100
+MAX_LOSS_CAP = 10
+WINDOW_FRACTION = 0.05
+
+
+@experimental_class('KillLossSpike')
+class KillLossSpike(Callback):
+    """Detects and handles loss spikes or high losses during training.
+
+    Monitors the training loss at the end of each batch and maintains a rolling window of recent losses.
+    If recent training losses exceed a specified cap or if a significant spike in loss is detected, the callback can either
+    log a warning (displayed as a message on the run event) or raise a LossSpikeError to stop the run without retry.
+
+    Args:
+        log_only (bool): If True, the callback will only log warnings without interrupting training. If False, a
+                         LossSpikeError will be raised to stop training upon detecting a loss spike or persistently
+                         high loss. Default is True.
+        patience (int): The number of consecutive outlier losses tolerated before considering the training loss to be
+                        persistently high. Default is 4 (so 5 consecutive outlier losses will trigger an error).
+        outlier_multiplier (int): The multiplier used to determine if a loss is an outlier. A loss is considered an
+                                  outlier if it is outlier_multiplier times greater than the mean of losses in
+                                  the current window. Default is 2.
+        window_size (int): The size of the rolling window used to track recent losses. This is set to 1/20 of the total training batches, with a minimum of 100 steps.
+        loss_cap (int): The maximum allowable loss. If the training loss consistently exceeds this value,
+                        it is considered a diverging or unstable run. This is set to the maximum loss from the first window of losses, with a maximum of 10.
+
+    Raises:
+        LossSpikeError: If log_only is False and a loss spike or persistently high loss is detected, this error is
+                        raised to stop the run with an error message.
+    """
+
+    def __init__(
+        self,
+        log_only: bool = True,
+        patience: int = 4,
+        outlier_multiplier: float = 2,
+    ):
+        self._enabled = True
```

Comment:
assuming the `_enabled` flag is still wip, but it should be set to `dist.get_global_rank() == 0` here, and then just used to gate all the calls to only happen if enabled

Reply:
So I tried this initially, but that meant the callback was not turned on at all (no logs + didn't stop when expected). Setting it in fit_start fixed this -- could be because the callback seems to be initialized before distributed training is started?

Reply:
hmm, that should not be the case.... it works fine for other things that we do it for: https://github.com/mosaicml/composer/blob/2fe9b10a0160466641aaf5890b5e865bc090fc78/composer/loggers/mlflow_logger.py#L151. The callback is created after dist has been initialized (and that function just reads from an env var anyway)

Reply:
Tried again and looks good now, must have been some issue with my run previously